CONVENTIONS.md

```diff
@@ -264,6 +264,9 @@ Please note that in case of [Single Node OpenShift](https://github.com/openshift
   * Two replicas
   * Hard pod [anti-affinity](https://kubernetes.io/docs/concepts/scheduling-eviction/assign-pod-node/#always-co-located-in-the-same-node) on hostname (no two pods on the same node)
   * Use the maxUnavailable rollout strategy on deployments (prefer 25% by default for a value)
+  * If the operator or operand requires persistent storage
+    * Consider that even with node-level anti-affinity, persistent storage can create a single point of failure by either provisioning volumes in the same availability zone or preventing some replicas from getting scheduled (e.g. the scheduler finds one good node permutation, but one node can't access sufficient free storage)
```

Comment:
> Consider that even with node-level anti-affinity, persistent storage can create a single point of failure

I wouldn't phrase it that way since even without persistent storage, node-level anti-affinity might still cause SPOF since nodes can be in the same AZ

The issue that we want to highlight here is when the `operator or operand with persistent storage` is migrated from soft anti-affinity to hard anti-affinity as it can cause scheduling issues. We could link the post mortem that was written in the past to detail the problem. In that case, the best we can recommend is to have the operator handle the rebalancing.

Reply:
> I wouldn't phrase it that way since even without persistent storage, node-level anti-affinity might still cause SPOF since nodes can be in the same AZ

Fair point, then intention here was to highlight that even with good pod-level affinity specs, persistent storage can create a SPOF. Will rephrase.

> The issue that we want to highlight here is when the `operator or operand with persistent storage` is migrated from soft anti-affinity to hard anti-affinity as it can cause scheduling issues. We could link the post mortem that was written in the past to detail the problem. In that case, the best we can recommend is to have the operator handle the rebalancing.

My understanding is that this is not just an artifact of moving from soft to hard anti-affinity. Even starting with a hard affinity spec can end up in the situation outlined above, can't it?

Reply:
Ah yes, you are correct. The issue happens when there is more than 1 replica before the migration to hard anti-affinity.

Reply:
> when the operator or operand with persistent storage is migrated from soft anti-affinity to hard anti-affinity as it can cause scheduling issues

From the scheduler POV this situation is specific to the volume plugins and how a new storage piece/claim is requested on a new node. Assuming there are two replicas with a persistent storage on the same node, after migrating to the hard anti-affinities, one of the persistent storage pieces will need to be deleted and created on a different node. I am unaware  of how this workflow is expected to proceed. Worth asking the storage folks for the details.

Reply:
I think it's nearly impossible to run an *operator* in HA with persistent storage. Below we require "an operator deployed by the CVO should run on control plane nodes". We typically run each master in a different availability zone and a storage volume can be usually available only in a single zone. With Deployment used for the operator and hard pod anti-affinity, the operator pods must run in different AZs and one of the replicas will not be able to mount the storage and hence it won't run at all. With StatefulSet, each replica will get its own volume and can run, but then they require some code in the operator to sync storage of each replica so both replicas have the same data. This code is usually not trivial.

I would strongly recommend *operators* not to require any PVs and store all their state in CustomResources. The API server will handle HA of this data.

For *operands*, the text makes sense. But it has the same issue - the operand must have some protocol to sync data between different PVs. Which is typical for databases running in some HA mode. In addition, depending on the underlying cloud, they might want to add anti-affinity rules to their pods to run each operand replica in a different zone. Such application could be tolerant to catastrophic failure of a single zone, on the other hand it would increase cross-zone traffic, which can be charged extra.

Reply:
>  With Deployment used for the operator and hard pod anti-affinity, the operator pods must run in different AZs and one of the replicas will not be able to mount the storage and hence it won't run at all

I guess with the exception of RWX volume types, this is mostly true. A RWX obviously implies that it has inbuilt mechanism to sync update between different highly available pods.

Reply:
@jsafrane You are correct on all point.
Prometheus has a somewhat rare HA setup where the replicas do not need synchronization, i.e. run with separate PVs. As mentioned in https://github.com/openshift/enhancements/pull/978#discussion_r781989100 we still wrote quite a bit of code to handle this when switching from soft to hard anti-affinity.

This PR intends to raise awareness about the impact that persistent storage can have on HA setups, _not_ to provide the solution.

Reply:
Therefore the guide should say that it's not recommended to use PVs for **operators**, AFAIK no OCP operator uses PVs (they may set up PVCs for the operands, of course)

For operands, the text is sort of OK and pod scheduling requirements (anti-affinity) and storage (availability zones, local storage) must be taken into account.

Reply:
Ahh yes fair point re. operators vs operands. I'll update accordingly, thank you.

Reply:
@jsafrane I updated the PR a bit, would appreciate if you could take another look.

Reply:
lgtm, I can see @simonpasquier found the same nits that I would report too.